mirgecom/logging_quantities.py

```diff
@@ -0,0 +1,306 @@
+
+    def set_sim_state(self, dim, state, eos) -> None:
+        """Update the state vector of the object."""
+        self.state_vars = extract_state_vars(dim, state, eos)
+
+# }}}
+
+# {{{ Discretization-based quantities
+
+
+class DiscretizationBasedQuantity(LogQuantity, StateConsumer):
+    """Logging support for physical quantities.
+
+    Possible rank aggregation operations (`op`) are: min, max, norm.
+    """
+
+    def __init__(self, discr: Discretization, quantity: str, op: str,
+                 unit: str = None, name: str = None, dim: int = None):
+        if unit is None:
+            unit = units_logging(quantity)
+
+        if name is None:
+            name = f"{op}_{quantity}" + (str(dim) if dim is not None else "")
+
+        LogQuantity.__init__(self, name, unit)
+        StateConsumer.__init__(self)
+
+        self.discr = discr
+
+        self.quantity = quantity
+        self.dim = dim
+
+        from functools import partial
+
+        if op not in ["min", "max", "norm"]:
+            raise ValueError("op must be one of 'min', 'max', 'norm'.")
+
+        if op == "min":
+            self._discr_reduction = partial(self.discr.nodal_min, "vol")
+            self.rank_aggr = min
+        elif op == "max":
+            self._discr_reduction = partial(self.discr.nodal_max, "vol")
+            self.rank_aggr = max
+        elif op == "norm":
+            self._discr_reduction = partial(self.discr.norm)
+            self.rank_aggr = max
+        else:
+            raise RuntimeError(f"unknown operation {op}")
+
+    @property
+    def default_aggregator(self):
+        """Rank aggregator to use."""
+        return self.rank_aggr
+
+    def __call__(self):
+        """Return the requested quantity."""
+        if self.state_vars is None:
+            return None
+
+        quantity = self.state_vars[self.quantity]
+
+        if self.dim is not None:  # momentum
+            return self._discr_reduction(quantity[self.dim])
+        else:  # mass, energy
+            return self._discr_reduction(quantity)
+
+# }}}
+
+
+# {{{ Kernel profile quantities
+
+class KernelProfile(MultiLogQuantity):
+    """Logging support for statistics of the OpenCL kernel profiling (time, \
+    num_calls, flops, bytes_accessed, footprint).
+
+    Parameters
+    ----------
+    actx
+        The array context from which to collect statistics. Must have profiling
+        enabled in the OpenCL command queue.
+
+    kernel_name
+        Name of the kernel to profile.
+    """
+
+    def __init__(self, actx: PyOpenCLArrayContext,
+                 kernel_name: str) -> None:
+        from mirgecom.profiling import PyOpenCLProfilingArrayContext
+        assert isinstance(actx, PyOpenCLProfilingArrayContext)
+
+        units = ["s", "GFlops", "1", "GByte", "GByte"]
+        names = [f"{kernel_name}_time", f"{kernel_name}_flops",
+                 f"{kernel_name}_num_calls", f"{kernel_name}_bytes_accessed",
+                 f"{kernel_name}_footprint"]
```

Comment:
Disagree with hardcoding this. Make the output of `get_profiling_data_for_kernel` a `dataclass`, and walk its fields.

Reply:
Done.